v2/call.js

```diff
@@ -100,19 +122,54 @@ CallRequest.RW.lazy.readService = function lazyReadService(frame) {
     return bufrw.str1.readFrom(frame.buffer, CallRequest.RW.lazy.serviceOffset);
 };
 
+CallRequest.RW.lazy.readServiceStr = function lazyReadServiceStr(frame) {
+    if (frame.cache.serviceStr !== null) {
+        return frame.cache.serviceStr;
+    }
+
+    var serviceStrEnd = null;
+    if (frame.cache.headerStartOffset) {
+        serviceStrEnd = frame.cache.headerStartOffset;
+    } else {
+        if (frame.size < CallRequest.RW.lazy.serviceOffset + 1) {
+            return null;
+        }
+        var strLength = frame.buffer.readUInt8(
+            CallRequest.RW.lazy.serviceOffset, false
+        );
+        serviceStrEnd = CallRequest.RW.lazy.serviceOffset + 1 + strLength;
+        frame.cache.headerStartOffset = serviceStrEnd;
+    }
+
+    if (frame.size < serviceStrEnd) {
+        return null;
+    }
+    var serviceNameStr = fastBufferToString(
+        frame.buffer,
+        CallRequest.RW.lazy.serviceOffset + 1,
+        serviceStrEnd
+    );
+
+    frame.cache.serviceStr = serviceNameStr;
+
+    return serviceNameStr;
+};
+
 CallRequest.RW.lazy.readHeaders = function readHeaders(frame) {
     // last fixed offset
     var offset = CallRequest.RW.lazy.serviceOffset;
 
-    // TODO: memoize computed offsets on frame between readService, readArg1,
-    // and any others
-
-    // SKIP service~1
-    var res = bufrw.str1.sizerw.readFrom(frame.buffer, offset);
-    if (res.err) {
-        return res;
+    if (frame.cache.headerStartOffset !== null) {
+        offset = frame.cache.headerStartOffset;
+    } else {
+        // SKIP service~1
+        var res = bufrw.str1.sizerw.readFrom(frame.buffer, offset);
```

Comment:
Can we avoid the `res` object alloc here? Why not just a standard `Buffer#readUint8` like above?